src/pipeline_v3/core/pipeline.py

```
@@ -63,13 +64,53 @@ async def fetch_document(source: str | Path) -> tuple[Path, str, bytes]:
     # Handle URL sources
     if isinstance(source, str) and (source.startswith(("http://", "https://"))):
+        # Validate URL for security
         try:
-            async with aiohttp.ClientSession() as session:
-                async with session.get(source) as response:
+            validated_url = URLSecurityValidator.validate_url(
+                source, allow_localhost=False, allow_private_ips=False
+            )
+        except SecurityError as e:
+            logger.error(f"URL security validation failed: {e}")
+            raise ValueError(f"URL validation failed: {e}")
```

Pylint (3.3.7)

[warning] Consider explicitly re-raising using 'raise ValueError(f'URL validation failed: {e}') from e'

(W0707)